src/rate_limiter.cc

```diff
@@ -775,52 +773,93 @@ RateLimiter::ResourceManager::AddModelInstance(
           resource.count();
     }
   }
+  // Increase max resource if needed.
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  UpdateMaxResource(pr.first->second);
+  return ParseAndValidateResources();
 }
 
 Status
 RateLimiter::ResourceManager::RemoveModelInstance(
     const ModelInstanceContext* instance)
 {
-  std::lock_guard<std::mutex> lk(model_resources_mtx_);
+  // Find instance from model resources.
+  std::lock_guard<std::mutex> lk1(model_resources_mtx_);
   const auto& itr = model_resources_.find(instance);
   if (itr == model_resources_.end()) {
     return Status(
         Status::Code::INTERNAL, "Can not find the instance to remove");
   }
+  // Check if max resources need to be updated.
+  bool update_needed = false;
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  for (const auto& resource_device_map : itr->second) {
+    auto ditr = max_resources_.find(resource_device_map.first);
+    if (ditr != max_resources_.end()) {
+      for (const auto& resource : resource_device_map.second) {
+        auto ritr = ditr->second.find(resource.first);
+        if (ritr != ditr->second.end() && ritr->second <= resource.second) {
+          update_needed = true;
+          if (ritr->second < resource.second) {
+            LOG_ERROR << "Should not print this! Removing an instance with "
```

Comment:
Why is this case treated as an error? Isn't it quite possible and is being handled below?

Reply:
> Isn't it quite possible and is being handled below?

Yes, this case [strictly less than] can be handled below.

> Why is this case treated as an error?

From the lifecycle of max resource as a whole, the max resource will always equal to the maximum count specified by one of the instance, for the particular resource identifier on the particular device. Therefore, given any instance and any given time, the resource requirement on the instance should not exceed the max resource. If this is seen [strictly less than case], it should be treated as an error.

The above is true even if explicit resource is set, because the max resource among instances is computed initially, and then the max resource is compared against explicit resource. If the explicit resource is less than the max resource, then [an error is returned](https://github.com/triton-inference-server/core/pull/208/files/eac9e0986569b776fc21cb3c28630e5a3b2e28d4#diff-bf3055556e25cd26a150f7262cf28d8e78733c10ed3f72f0c4fd10d9fe010dbbR949-R957). If the explicit resource equals to or exceeds the max resource, then the [max resource is updated to match the explicit resource](https://github.com/triton-inference-server/core/pull/208/files/eac9e0986569b776fc21cb3c28630e5a3b2e28d4#diff-bf3055556e25cd26a150f7262cf28d8e78733c10ed3f72f0c4fd10d9fe010dbbL920). From the logic, the max resource can only be increased further, and cannot be lowered. Therefore, it would be very unusual to see an instance to have resource count exceeding the max resource.

Reply:
Thanks for the explanation!